sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+
+    @Nullable abstract ConnectionConfiguration getConnectionConfiguration();
+    abstract long getBatchSize();
+    abstract int getBatchSizeMegaBytes();
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration connectionConfiguration);
+      abstract Builder setBatchSize(long batchSize);
+      abstract Builder setBatchSizeMegaBytes(int batchSizeMegaBytes);
+      abstract Write build();
+    }
+
+    public Write withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "ElasticsearchIO.write()"
+          + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Write withBatchSize(long batchSize) {
+      return builder().setBatchSize(batchSize).build();
+    }
+
+    public Write withBatchSizeMegaBytes(int batchSizeMegaBytes) {
+      return builder().setBatchSizeMegaBytes(batchSizeMegaBytes).build();
+    }
+
+    @Override
+    public void validate(PCollection<String> input) {
+      checkState(getConnectionConfiguration() != null, "ElasticsearchIO.write() requires a "
+          + "connection configuration to be set via withConnectionConfiguration(configuration)");
+    }
+
+    @Override
+    public PDone apply(PCollection<String> input) {
+      input.apply(ParDo.of(new WriterFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    private static class WriterFn extends DoFn<String, Void> {
+
+      private Write spec;
+
+      private transient RestClient restClient;
+      private ArrayList<String> batch;
+      private long currentBatchSizeBytes;
+
+      public WriterFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void createClient() throws Exception {
+        if (restClient == null) {
+          restClient = spec.getConnectionConfiguration().createClient();
+        }
+      }
+
+      @StartBundle
+      public void startBundle(Context context) throws Exception {
+        batch = new ArrayList<>();
+        currentBatchSizeBytes = 0;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        String document = context.element();
+        batch.add(String.format("%s%n%s%n", "{ \"index\" : {} }", document));
+        currentBatchSizeBytes += document.getBytes().length;
+        if (batch.size() >= spec.getBatchSize()
+            || currentBatchSizeBytes >= (spec.getBatchSizeMegaBytes() * 1024L * 1024L)) {
+          finishBundle(context);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle(Context context) throws Exception {
+        StringBuilder bulkRequest = new StringBuilder();
+        if (batch.size() > 0) {
```

Comment:
I mean transform code into `if (batch.isEmpty()) return;` and reduce nesting of lines 587-621, so the code will be easier to read - e.g. the current code kept me holding my breath for "is there an else branch, or anything after the if?"